### PR TITLE
remove "default" CaseFormat for stringConstants!

### PR DESCRIPTION
**Pull request by dglachs**
this solves the following problem:

when having similar names for attributes and classes, such as
- ssn:FeatureOfInterest (denoting a class)
- ssn:featureOfInterest (denoting a property)

having UPPER_UNDERSCORE as the default stringConstantCase will cause duplicate fields ...and there is no possibility to omit reformatting to any of the CaseFormat options.

### CODE REVIEW

**Comment by ansell** [comment]
Are you able to modify the maven property to do what you need without changing the default? 

The reason that the default case format exists is to make it work for the majority of cases that we need it to work for, while making sure that others are able to reconfigure it themselves in their pom.xml files.

**Comment by dglachs** [comment]
Hello Peter, Thomas,

the problem I see, that there is no valid option in com.google.common.base.CaseFormat which says "leave as it is". In fact, for now, there are the following options:

1. omitting the tag <stringConstantCase>: this will do the default of UPPER_UNDERSCORE 
2. provide the tag but with no value: this will do the default of UPPER_UNDERSCORE
3. use another valid CaseFormat option (LOWER_HYPHEN, UPPER_HYPHEN etc.): this will do the requested case adjustments but result in a duplicate field (at least for the SSN ontology as mentioned in my request and also for the FOAF ontology where for example foaf:familyName and foaf:family_name cause a conflict)
4. use an invalid option such as DO_NOTHING: this will result in an error since the configuration cannot be parsed.

For me, the simplest solution was to remove the default, because this results in a "no case format adjustment" when omitting the <stringConstantCase> tag. 

However, a more sophisticated solution could allow
- specify the default generation options for URI's and/or strings including case format, prefix, suffix 
- allow overriding the default settings for each configured vocabulary (partially already done with vocabulary-specific caseFormat) 

Best
  Dietmar

**Comment by ansell** [comment]
Thanks, that explains it quite well I think. In that case, I would definitely support offering either a dummy value or default to nothing and the current users update our pom.xml files with our desired cases.

**Comment by dglachs** [comment]
there was some time left so I modified the functionality quite a bit in order to provide more (generation) flexibility: 

(1) allow to specify prefix & suffix for the URI constants as well. 
either provide the tags in the plugin as 

    <createUriConstants>true</createUriConstants>    (defaults to true)
    <uriConstantPrefix></uriConstantPrefix> (defaults to "")
    <uriConstantSuffix></uriConstantSuffix> (defaults to "")
    <createStringConstants>true</createStringConstants>    (defaults to true)
    <stringConstantPrefix></stringConstantPrefix> (defaults to "")
    <stringConstantSuffix></stringConstantSuffix> (defaults to "_STRING")

or in an special "setting" structure (which may also be used for each vocabulary)

    <uriGeneration>   (or <stringGeneration>)
    ...<caseFormat></caseFormat>
    ...<constantPrefix></constantPrefix>
    ...<constantSuffix></constantSuffix> 
    </uriGeneration>

this for example gives you the possibility to configure a "_URI" suffix for URI's (if desired) . in addition, vocabulary specific settings are also possible. 

(2) default case format will be used for both URI & String generation...
I personally found it not logical to have URI constants adjusted differentlty than String constants by default. So I'd propose to have only one "caseFormat" element which is used for both generation options by default. now, no caseFormat adjustment for URI's takes place by default (as it was before). the caseFormat setting is then used for string generation as well, unless it is further specified in a &lt;stringGeneration&gt; element.

(3) no changes in pom.xml of your users required, command line parameters are still the same
the changes are just additions, nothing has been removed.

finally, there were quite some adjustments and changes in the maven plugin and in the VocabBuilder class as well. Maybe you find the changes useful.

**Comment by ja-fra** [comment]
I'd suggest to switch to the new configuration format @dglachs suggests and deprecate the old way (to be removed in one the next versions). Keeping both is too much overhead and will confuse users.

For the defaults: I'd add a special handling for `AS_IS` or `NONE` which then should be also used as default/not explicitly set...

And finally: After merging these changes we should have new release!

**Comment by tkurz** [comment]
I totally agree with @ja-fra . @dglachs could you add the 'NONE' case? Afterwards we can merge and release.

**Comment by dglachs** [comment]
Added the required functionality ... 

currently the following tags in the &lt;configuration&gt; section are possible:

```
<caseFormat>NONE</caseFormat>
```

this is the default and specifies NO caseFormat-Adjustments for both URI & String-Constant generation. (the distinct default generation options are still specified with the "old" parameters, see my previous comment). For having more fine grained generation options - the GenerationSetting class is used. One may specify

```
<uriGeneration>   (or <stringGeneration>)
    <caseFormat></caseFormat>
    <constantPrefix></constantPrefix>
    <constantSuffix></constantSuffix> 
</uriGeneration>
```

which then override the default-settings. to have it more sophisticated it is also possible to have the following structure along with a vocabulary (under &lt;vocabularies&gt;)

```
<vocabulary>
    <className>FOAF</className>
    <prefix>http://xmlns.com/foaf/0.1/</prefix>
    <file>${project.build.directory}/../ontologies/foaf.ttl</file>
    <uriGeneration>   (or <stringGeneration>)
        <caseFormat></caseFormat>
        <constantPrefix></constantPrefix>
        <constantSuffix></constantSuffix> 
    </uriGeneration>
</vocabulary>
```

for specifying specific prefix/suffix for a particular vocabulary (overriding the general settings from above). 

As noted - there are still the "duplicated" or "old" parameters (&lt;stringConstantPrefix&gt;,&lt;stringConstantSuffix&gt; etc.) - I've marked them as deprecated. Anyways, as @ja-fra suggested, these "old" parameters should be removed, for this, it is recommended to remove only the Parameter-Annotation and just leave the variables including their default as constants.

File: vocab-builder-cli/src/main/java/com/github/tkurz/sesame/vocab/Main.java [hunk 1]
```diff
@@ -100,38 +100,82 @@ public static void main(String[] args) {
...
-            if (cli.hasOption('P')) {
-                builder.setStringPropertyPrefix(cli.getOptionValue('P'));
-            } else {
-                builder.setStringPropertyPrefix(null);
-            }
+            CaseFormat caseFormat = null;
+            final GenerationSetting uriGenerationSetting = new GenerationSetting(); 
             if (cli.hasOption('c')) {
+            	String uriCaseFormat = cli.getOptionValue('c');
+            	uriGenerationSetting.setCaseFormat(uriCaseFormat);
```
Comment (ansell) [inline vocab-builder-cli/src/main/java/com/github/tkurz/sesame/vocab/Main.java]: This is an odd style of programming to optimistically set the case format and then check it later and throw an exception. Typically, to make code maintainable, the setter is only called after it is validated (or the validation happens inside of the setter).

File: vocab-builder-cli/src/main/java/com/github/tkurz/sesame/vocab/Main.java [hunk 2]
```diff
@@ -100,38 +100,82 @@ public static void main(String[] args) {
...
-                builder.setStringPropertySuffix(cli.getOptionValue('S'));
-            } else {
-                builder.setStringPropertySuffix(null);
-            }
-            if (cli.hasOption('P')) {
-                builder.setStringPropertyPrefix(cli.getOptionValue('P'));
-            } else {
-                builder.setStringPropertyPrefix(null);
-            }
+            CaseFormat caseFormat = null;
```
Comment (ansell) [inline vocab-builder-cli/src/main/java/com/github/tkurz/sesame/vocab/Main.java:103]: This does not need to be a shared variable between "c" and "C" parsing. It doesn't even look like it needs to be stored, as it is only used for a null check in both locations and the string is what is passed into GenerationSetting now instead of the CaseFormat object.

Is it redundant to pass the string around instead of passing a CaseFormat object around?

File: vocab-builder-cli/src/main/java/com/github/tkurz/sesame/vocab/Main.java [hunk 3]
```diff
@@ -100,38 +100,82 @@ public static void main(String[] args) {
...
+            	if ( cli.hasOption('P')) {
+            		setting.setConstantPrefix(cli.getOptionValue('P'));
+            	}
+            	if ( cli.hasOption('S')) {
+            		setting.setConstantSuffix(cli.getOptionValue('S'));
+            	}
+                if (cli.hasOption('C')) {
+                	setting.setCaseFormat(cli.getOptionValue('C'));
+                    try {
+                        final CaseFormat strCaseFormat = CaseFormat.valueOf(cli.getOptionValue('C'));
```
Comment (ansell) [inline vocab-builder-cli/src/main/java/com/github/tkurz/sesame/vocab/Main.java]: Ditto comment above.

File: vocab-builder-cli/src/main/java/com/github/tkurz/sesame/vocab/Main.java [hunk 4]
```diff
@@ -100,38 +100,82 @@ public static void main(String[] args) {
...
+                        if (caseFormat == null) {
```
Comment (ansell) [inline vocab-builder-cli/src/main/java/com/github/tkurz/sesame/vocab/Main.java]: This looks like it is checking the wrong variable, as a result of pushing the caseFormat variable into the main method scope.

File: vocab-builder-cli/src/main/java/com/github/tkurz/sesame/vocab/Main.java [hunk 5]
```diff
@@ -100,38 +100,82 @@ public static void main(String[] args) {
...
-                    throw new ParseException("Did not recognise constantCase: Must be one of " + Arrays.asList(CaseFormat.values()));
                 }
+                else {
+                	// uri case format applies as DEFAULT
+                	setting.setCaseFormat("NONE");
+                }
+                // specify string generation
+            	builder.setStringGeneration(setting);
             }
+//            if (cli.hasOption('S')) {
```
Comment (ansell) [inline vocab-builder-cli/src/main/java/com/github/tkurz/sesame/vocab/Main.java]: Just delete the following code rather than commenting it. It is always available if we need it in the git repository.

File: vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/GenerationSetting.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import com.google.common.base.CaseFormat;
+
+public class GenerationSetting {
+
+	private String caseFormat;
+	private String constantPrefix;
+	private String constantSuffix;
+	public CaseFormat getCaseFormat() {
+		try {
+			if ( caseFormat != null ) {
```
Comment (ansell) [inline vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/GenerationSetting.java]: No need for this null check if we are catching all exceptions. CaseFormat.valueOf will throw an NPE for that.

Alternatively, the null check could be outside of the try-catch block to emphasise that the exceptions are expected from CaseFormat.valueOf

File: vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/GenerationSetting.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+				return CaseFormat.valueOf(caseFormat);
+			}
+		} catch (Exception e) {
+		}
+		return null;
+	}
+	public void setCaseFormat(String caseFormat) {
+		this.caseFormat = caseFormat;
```
Comment (ansell) [inline vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/GenerationSetting.java]: It would be useful to do validation here rather than waiting until the get method is called to validate the input.

File: vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/GenerationSetting.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+	public static GenerationSetting createDefault(CaseFormat format, String prefix, String suffix) {
+		GenerationSetting setting = new GenerationSetting();
+		setting.caseFormat = (format != null ? format.name() : null);
+		setting.constantPrefix = prefix;
+		setting.constantSuffix = suffix;
+		return setting;
+		
+	}
+	
+	public GenerationSetting mergeWith(GenerationSetting other) {
```
Comment (ansell) [inline vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/GenerationSetting.java]: Need javadoc to specify the contract here, as it isn't clear what semantics the merge operation has without reading the code.

**Comment by ansell** [comment]
Overall, the code also needs to be reformatted to use spaces instead of tabs.

File: vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/VocabBuilder.java [hunk 9]
```diff
@@ -190,73 +197,150 @@ public void generate(String className, PrintWriter out) throws IOException, Grap
...
+						String.format("%s%s%s", StringUtils.defaultString(uriGeneration.getConstantPrefix()),
+								doCaseFormatting(key, uriGeneration.getCaseFormat()),
+								StringUtils.defaultString(uriGeneration.getConstantSuffix())));
+	            out.printf(getIndent(2) + "%s = factory.createURI(%s.NAMESPACE, \"%s\");%n", nextKey, className, key);
+	        }
+	        out.println(getIndent(1) + "}");
+	        out.println();
+
+		}
+        //string constant values
```
Comment (ansell) [inline vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/VocabBuilder.java]: Delete the following commented out code, and after reformatting the code above to use spaces, git will show what the changes you made to this section are.

File: vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/VocabBuilder.java [hunk 10]
```diff
@@ -296,6 +380,10 @@ public void generateResourceBundle(String baseName, Path bundleDir) throws Gener
     }
 
     public HashMap<String, Properties> generateResourceBundle(String baseName) throws GenerationException {
+    	// be sure to have at least the default URI constant settings
+    	if ( uriGeneration == null ) {
```
Comment (ansell) [inline vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/VocabBuilder.java:321]: The code style we have used here so far is not to have spaces inside of the if statement brackets. Ie, change to:

```
if (uriGeneration == null) {
```

File: vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/VocabBuilder.java [hunk 11]
```diff
@@ -317,7 +405,8 @@ public void generateResourceBundle(String baseName, Path bundleDir) throws Gener
         bundles.put(baseName, new Properties());
         for (String key : keys) {
             final URI resource = splitUris.get(key);
-            String nextKey = cleanKey(doCaseFormatting(key, getConstantCase()));
+            //
```
Comment (ansell) [inline vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/VocabBuilder.java:345]: Was this meant to have a comment on this line?

File: vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/VocabBuilder.java [hunk 12]
```diff
@@ -485,34 +574,51 @@ public String getPreferredLanguage() {
...
 
     public void setConstantCase(CaseFormat caseFormat) {
-        this.caseFormat = caseFormat;
+    	this.caseFormat = caseFormat;
     }
 
     public CaseFormat getConstantCase() {
         return caseFormat;
     }
+//
```
Comment (ansell) [inline vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/VocabBuilder.java:520]: As above, delete code instead of commenting it out.

File: vocab-builder-maven-plugin/src/main/java/com/github/tkurz/sesame/vocab/plugin/VocabularyBuilderMojo.java [hunk 13]
```diff
@@ -245,19 +313,50 @@ public void execute() throws MojoExecutionException, MojoFailureException {
                         target = target.resolve(builder.getPackageName().replaceAll("\\.", "/"));
                         Files.createDirectories(target);
                     }
+                    if ( createUriConstants && uriGeneration != null ) {
```
Comment (ansell) [inline vocab-builder-maven-plugin/src/main/java/com/github/tkurz/sesame/vocab/plugin/VocabularyBuilderMojo.java]: This check negates the safety above to "be sure to generate at least the URI's".

Comment (ansell) [inline vocab-builder-maven-plugin/src/main/java/com/github/tkurz/sesame/vocab/plugin/VocabularyBuilderMojo.java]: Sorry, should have read down further first.

Comment (ansell) [inline vocab-builder-maven-plugin/src/main/java/com/github/tkurz/sesame/vocab/plugin/VocabularyBuilderMojo.java]: Nope, it does actually negate the global safety. You need to reorganise this code and document why the merge would not happen with vocab, but it would happen in this case. Naively, although it is hard to follow your code, I would have thought you would always want to merge the global settings into the vocab specific settings.

File: vocab-builder-maven-plugin/src/main/java/com/github/tkurz/sesame/vocab/plugin/VocabularyBuilderMojo.java [hunk 14]
```diff
@@ -245,19 +313,50 @@ public void execute() throws MojoExecutionException, MojoFailureException {
...
+                    	builder.setStringGeneration(stringGeneration.mergeWith(vocab.getStringGeneration()));
                     }
+                    else if (vocab.getStringGeneration() != null ) { 
+                    	/*
+                    	 * when string constant generation deactivated, vocabulary specific settings
+                    	 * may apply 
+                    	 */
+                    	builder.setStringGeneration(vocab.getStringGeneration());
+                    }
+//                    if ( uriGeneration != null ) {
```
Comment (ansell) [inline vocab-builder-maven-plugin/src/main/java/com/github/tkurz/sesame/vocab/plugin/VocabularyBuilderMojo.java]: Delete code instead of commenting it out so git can show the full changes

File: vocab-builder-maven-plugin/src/main/java/com/github/tkurz/sesame/vocab/plugin/VocabularyBuilderMojo.java [hunk 15]
```diff
@@ -314,6 +413,19 @@ public void execute() throws MojoExecutionException, MojoFailureException {
             throw new MojoExecutionException("Could not write Vocabularies", e);
         }
     }
+    /**
+     * Helper method for initializing based on the provided caseFormat string 
+     * @param caseFormat
+     * @return
+     */
+    private CaseFormat caseFormatFromString(String caseFormat) {
```
Comment (ansell) [inline vocab-builder-maven-plugin/src/main/java/com/github/tkurz/sesame/vocab/plugin/VocabularyBuilderMojo.java]: This code should not be replicated both here and in GenerationSetting. Need to have a static method in a helper class so that we ensure that we are doing the same operations everywhere for this case.

File: vocab-builder-maven-plugin/src/main/java/com/github/tkurz/sesame/vocab/plugin/VocabularyBuilderMojo.java (hunk 13 above)
Comment (dglachs) [inline vocab-builder-maven-plugin/src/main/java/com/github/tkurz/sesame/vocab/plugin/VocabularyBuilderMojo.java]: No, the global settings will be overridden with the vocabulary specific configuration - if set.

File: vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/VocabBuilder.java (hunk 10 above)
Comment (dglachs) [inline vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/VocabBuilder.java]: ups

File: vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/GenerationSetting.java (hunk 7 above)
Comment (dglachs) [inline vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/GenerationSetting.java]: yes - will change it

**Comment by dglachs** [comment]
reverted the NONE option commit since it created more problems than it solved ...

in fact the NONE option was present before because simply omitting the caseFormat-tag (at any level) was the NONE option.  

anyways the proposed configuration structure is still valid

**Comment by tkurz** [comment]
The code seems to be okay now. Should we merge the request and release a new version?

**Comment by ansell** [comment]
Yes, the functionality looks good, just some style issues remaining that we can automatically reformat after merging and delete some of the commented code that is still present.

Can we switch on TravisCI to do automated tests btw? I thought it was on because we have the .travis.yml file, but it doesn't seem to be enabled at the other end right now:

https://travis-ci.org/tkurz/sesame-vocab-builder
